lib/rust/parser/src/syntax/tree.rs

```diff
@@ -231,6 +396,192 @@ impl<'s> span::Builder<'s> for MultiSegmentAppSegment<'s> {
 
 
 
+// ====================
+// === BlockBuilder ===
+// ====================
+
+/// Builds an AST block type from a sequence of lines.
+///
+/// Note that the block type is not fully determined at this stage: We apply context information
+/// later (see `apply_operator`) to distinguish the two non-operator block types, `BodyBlock` and
+/// `ArgumentBlockApplication`. Here we treat every non-operator block as an argument block,
+/// because creating a body block involves re-interpreting the expressions in statement context.
+///
+/// The implementation is a state machine. The only top-level transitions are:
+/// - `Indeterminate` -> `Operator`
+/// - `Indeterminate` -> `NonOperator`
+///
+/// The `Operator` state has two substates, and one possible transition:
+/// - `body_lines is empty` -> `body_lines is not empty`
+#[derive(Debug)]
+pub enum BlockBuilder<'s> {
+    /// The builder is in an indeterminate state until a non-empty line has been encountered, which
+    /// would distinguish an operator-block from a non-operator block.
+    Indeterminate {
+        /// The `Newline` token introducing the block, and `Newline` tokens for any empty lines
+        /// that have been encountered.
+        empty_lines: Vec<token::Newline<'s>>,
+    },
+    /// Building an operator block. If any line doesn't fit the operator-block syntax, that line
+    /// and all following will be placed in `body_lines`.
```

Comment:
these block-related things are big - can we refactor all of them to `tree::block` please?